### PR TITLE
Add Progressive Web App (PWA) functionality

### PR DESCRIPTION
**Pull request by caleb531**
- Add a web app manifest
- Add service worker for offline access

With these two features in place, Yugioh Calculator will be installable as a native app on Chrome and Android. The installed app will also utilize the icon and defined colors to present a splash screen while loading.

Yugioh Calculator now scores a 93/100 on the PWA reporting tool, Lighthouse. The JSON for the above report is attached below. To view it, drag it into the [online viewer](https://googlechrome.github.io/lighthouse/viewer/).

[localhost_2017-05-15_20-21-41.json.zip](https://github.com/jacksonrayhamilton/yugioh-calculator-2016/files/1003124/localhost_2017-05-15_20-21-41.json.zip)

screen shot 2017-05-15 at 8 30 04 pm

Currently, the service worker does not handle Google Analytics. However, there *is* [such a package](https://developers.google.com/web/updates/2016/07/offline-google-analytics) for offline GA. I will explore this option soon.

### CODE REVIEW

File: Gruntfile.js
Comment (jacksonrayhamilton) [inline Gruntfile.js:24]: This breaks favicon generation.  `icon.16x16.png` is no longer used for the generated `favicon.ico`.

That said, comparing `icon.16x16.png` and the generated imagemagick icon, I think the quality of the latter is good enough.  So can you delete `icon.16x16.png` please?

File: app/app.js
Comment (jacksonrayhamilton) [inline app/app.js]: End explanatory fragments with a period, please.

File: app/assets/index.html
Comment (jacksonrayhamilton) [inline app/assets/index.html]: Inconsistent space before `/>`.

File: app/assets/index.html
Comment (jacksonrayhamilton) [inline app/assets/index.html]: Can you use a CSS class and add more styling here?

A "reload" button might also be nice (i.e. an HTML link pointing to the same page).

I am thinking that a light gray rectangle with rounded corners (like the one used for life points) would be in the top part of the screen instructing to enable JS, and a dark gray rectangle with rounded corners (like the one used for digits) would be in the lower part of the screen and provide the "reload" button.

[General idea.](https://shareitdammit.com/MIoJiIfm)  Feel free to improve or be more creative.

**Review by jacksonrayhamilton — CHANGES_REQUESTED** [review]
A few things that I'd like addressed but do feel free to oppose me.

Also I'll wait until the GA thing to merge.

File: Gruntfile.js
Comment (caleb531) [inline Gruntfile.js]: Darn, I forgot to update the icon path in `Gruntfile.js`—do you still want me to do that, or is it irrelevant now that you are declaring `icon-16x16.png` obsolete?

File: app/assets/index.html
Comment (caleb531) [reply app/assets/index.html]: A "reload" button won't work if JS is disabled, but yes, I'll add prettier and more-consistent styling.

Comment (jacksonrayhamilton) [reply app/assets/index.html]: This wouldn't work with JS disabled?

```html
<a class="button" href="/">Reload</a>
```

File: app/styles/theme.css
Comment (jacksonrayhamilton) [inline app/styles/theme.css]: The purpose of `theme.css` is to apply styling like colors and spacing all in one place, and this centralization is supposed to help keep that styling consistent.  Part of this is due to the decision to use plain CSS, which is half tied up with the project's history, and half as an experiment.

Anyway, properties declared here that you see declared elsewhere in this file should stay here, and probably the rest should be moved to a file called `noscript.css`.

You may also notice that for essentially all other styles where it was possible, I opted to use classes instead of element selectors.  I believe this is a more reliable way of writing CSS for a web app, and would prefer if these styles could be declared under class selectors instead.

File: Gruntfile.js
Comment (caleb531) [reply Gruntfile.js:24]: Done! :)

**Comment by jacksonrayhamilton** [comment]
Two issues with the noscript styling:

1) `<div class="yc-app"></div>` is still on the page, which doubles the vertical space and creates scrollbars.

2) I'd prefer if the rectangles were the same width.

**Comment by caleb531** [comment]
Done! Also, feel free to merge this PR whenever you're ready (or perhaps once you've performed your extensive benchmarking!).

**Comment by caleb531** [comment]
And... I forgot to address your point about the rectangles being of equal width. Stand by for a commit to fix that...

**Comment by caleb531** [comment]
Okay, *now* it is finished! Enjoy your new Progressive Yugioh Calculator! :)

**Comment by jacksonrayhamilton** [comment]
The styling leaves a bit to be desired on what Chrome claims to be the iPad screen resolution: https://shareitdammit.com/qio77uFO  Will you ensure it looks pleasing on all common screen resolutions?

File: app/styles/noscript.css
Comment (jacksonrayhamilton) [inline app/styles/noscript.css]: All other classes in this app are prefixed with "yc-", so these classes should start with "yc-no-js-", instead.

File: app/styles/noscript.css
Comment (jacksonrayhamilton) [inline app/styles/noscript.css]: Missing space before closing `*/`.

**Review by jacksonrayhamilton — CHANGES_REQUESTED** [review]
More trivial updates.

**Comment by jacksonrayhamilton** [comment]
I was able to test this with https://github.com/caleb531/yugioh-calculator-2016/pull/3.  Indeed, the app now works offline on Android, both in Android Chrome and when added to the home screen (which I was immediately prompted to do at https://jacksonrayhamilton.github.io/yugioh-calculator-2016/).

Thanks for your work on this so far!

**Comment by caleb531** [comment]
Alright, missing spaces added, classes prefixed, and noscript text overflow fixed:

screen shot 2017-06-10 at 11 13 47 am

screen shot 2017-06-10 at 11 13 40 am

**Review by jacksonrayhamilton — APPROVED** [review]

**Comment by jacksonrayhamilton** [comment]
It is done.

**Pull request merged**
